usm: ssl: openssl: Support SSL_read_ex and SSL_write_ex methods

guyarb opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

Adding support with hooking `SSL_read_ex` and `SSL_write_ex`.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succint form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation
`SSL_read_ex` and `SSL_write_ex` methods were introduced in openssl 1.1 and started being used by some applications. Starting from python 3.10, those are the methods being used by default to maintain a TLS connection.
However, those methods might not exist in any version, so we are trying to do best effort
<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

guyarb committed:
usm: ssl: openssl: Support SSL_read_ex and SSL_write_ex methods

guyarb commented:
Still a draft PR
Blocked on:
- https://github.com/DataDog/ebpf-manager/pull/64
- Adding kitchen tests

nplanel commented on pkg/network/http/ebpf_ssl.go:
```diff
@@ -198,37 +358,38 @@ func (o *sslProgram) Stop() {
+				p, found := m.GetProbe(identifier)
+				if found {
+					if !p.IsRunning() {
+						err := p.Attach()
```
sowatcher would probably need some change if SSL_read/write_ex() doesn't exist, can't be binded (ex: openssl 1.0)
But this could be addressed in different PR

guyarb replied:
I'm using the Selectors and marked `SSL_{read, write}_ex` methods as "BestEffort" for that purpose

guyarb replied:
resolving as we spoke about that in person

guyarb commented:
rebased to pull main changed

guyarb committed:
usm: ssl: openssl: Added tracer tests

The test will setup an https python server, perform couple of requests and make sure all were captured.
The CI kitchen tests will run that test on various of machines with different python versions (3.6 to 3.10)
so at the moment, it will cover all scenarios of python versions

DanielLavie commented on pkg/ebpf/bytecode/runtime/http.go:
```diff
@@ -4,4 +4,4 @@
 package runtime
 
-var Http = NewAsset("http.c", "9725a8188d39fba30e5d5f02ba30fd50b92e733f971aeeedc2dcd0ea66dc7ba0")
+var Http = NewAsset("http.c", "2b98b84c2e5afafb2a527cb95a2c5438bacc7ac921e30eaf1b9b6828662b6dbc")
```
Can't we use a build tag to include this hash?

guyarb committed:
usm: ssl: openssl: Added missing imports

DanielLavie commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -225,6 +225,107 @@ int uretprobe__SSL_write(struct pt_regs* ctx) {
+    args.buf = (void *)PT_REGS_PARM2(ctx);
+    args.size_out_param = (size_t *)PT_REGS_PARM4(ctx);
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    log_debug("uprobe/SSL_read_ex: pid_tgid=%llx ctx=%llx\n", pid_tgid, args.ctx);
```
Would is be helpful to log the ctx pointer?

guyarb replied:
if you ask me, no
but i stick with the "convention" around here
we should revisit it a different discussion and PR

DanielLavie replied:
Can you open a task for it?

guyarb replied:
https://datadoghq.atlassian.net/browse/USMO-111

DanielLavie commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -225,6 +225,107 @@ int uretprobe__SSL_write(struct pt_regs* ctx) {
+    args.ctx = (void *)PT_REGS_PARM1(ctx);
+    args.buf = (void *)PT_REGS_PARM2(ctx);
+    args.size_out_param = (size_t *)PT_REGS_PARM4(ctx);
+    u64 pid_tgid = bpf_get_current_pid_tgid();
```
const

guyarb replied:
cannot be const, as it being passed a pointer

DanielLavie commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -225,6 +225,107 @@ int uretprobe__SSL_write(struct pt_regs* ctx) {
+
+SEC("uretprobe/SSL_read_ex")
+int uretprobe__SSL_read_ex(struct pt_regs* ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
```
const

DanielLavie commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -225,6 +225,107 @@ int uretprobe__SSL_write(struct pt_regs* ctx) {
+SEC("uretprobe/SSL_read_ex")
+int uretprobe__SSL_read_ex(struct pt_regs* ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    int return_code = (int)PT_REGS_RC(ctx);
```
Why not check the return code as first step? do we really need this log message?

guyarb replied:
it is not for the log, but for the cleanup

DanielLavie commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -225,6 +225,107 @@ int uretprobe__SSL_write(struct pt_regs* ctx) {
+    args.size_out_param = (size_t *)PT_REGS_PARM4(ctx);
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    log_debug("uprobe/SSL_read_ex: pid_tgid=%llx ctx=%llx\n", pid_tgid, args.ctx);
+    bpf_map_update_elem(&ssl_read_ex_args, &pid_tgid, &args, BPF_ANY);
```
What about the return value? don't we have telemetry for that?

guyarb replied:
of `bpf_map_update_elem`?

DanielLavie replied:
True

guyarb replied:
irrelevant at the moment
it is best effort

usamasaqib replied:
If you wish you can begin collecting telemetry for this map, after rebasing with main, by replacing the helper call with `bpf_map_update_with_telemetry`.

DanielLavie commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -225,6 +225,107 @@ int uretprobe__SSL_write(struct pt_regs* ctx) {
+int uretprobe__SSL_read_ex(struct pt_regs* ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    int return_code = (int)PT_REGS_RC(ctx);
+    if (return_code <= 0) {
```
From https://www.openssl.org/docs/man1.1.1/man3/SSL_read_ex.html:
`SSL_read_ex() and SSL_peek_ex() will return 1 for success or 0 for failure`
I would check `return_code != 1`

DanielLavie commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -225,6 +225,107 @@ int uretprobe__SSL_write(struct pt_regs* ctx) {
+
+    https_process(t, args->buf, bytes_count, LIBSSL);
+cleanup:
+    bpf_map_delete_elem(&ssl_read_ex_args, &pid_tgid);
```
Do we really need to explicitly delete it? the next time it will be called it should be overwritten with the new value. I'm not totally against it, I just think the cleanup pattern can cause some problems and is creating some weird flows in the code (like not going to cleanup when `args == NULL`)

guyarb replied:
we want to clean irrelevant entries from the memory and reduce the memory utilization by our hooks

DanielLavie replied:
The map is pre-allocated with fix size, are you sure it reduces the size?

guyarb replied:
yes

DanielLavie commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -225,6 +225,107 @@ int uretprobe__SSL_write(struct pt_regs* ctx) {
+    size_t bytes_count = 0;
+    bpf_probe_read_user(&bytes_count, sizeof(bytes_count), args->size_out_param);
+    if ( bytes_count <= 0) {
+        log_debug("uretprobe/SSL_read_ex: pid_tgid=%llx len=%d\n", pid_tgid, bytes_count);
```
Can this message state the problem more clearly?

DanielLavie commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -225,6 +225,107 @@ int uretprobe__SSL_write(struct pt_regs* ctx) {
+    }
+
+    void *ssl_ctx = args->ctx;
+    conn_tuple_t *t = tup_from_ssl_ctx(ssl_ctx, pid_tgid);
```
Consider using clearer variable names than single letters

DanielLavie commented on pkg/network/http/ebpf_ssl.go:
```diff
@@ -24,37 +24,197 @@ import (
-	"uprobe/SSL_write":           "uprobe__SSL_write",
-	"uretprobe/SSL_write":        "uretprobe__SSL_write",
-	"uprobe/SSL_shutdown":        "uprobe__SSL_shutdown",
+var openSSLProbes = []manager.ProbesSelector{
```
Nice :)

DanielLavie commented on pkg/network/http/ebpf_ssl.go:
```diff
@@ -198,37 +358,37 @@ func (o *sslProgram) Stop() {
-						return err
-					}
+
+		for i := range probes {
```
```suggestion
		for singleProbe := range probes {
```

guyarb replied:
`i` is the index

DanielLavie replied:
You are right, can't we operate on the single probe? why do we need the index?

guyarb replied:
changed

DanielLavie commented on pkg/network/http/ebpf_ssl.go:
```diff
@@ -198,37 +358,37 @@ func (o *sslProgram) Stop() {
+					UID:          uid,
 				}
+				probes[i].EditProbeIdentificationPair(selector, identifier)
+				p, found := m.GetProbe(identifier)
```
What is p?

guyarb replied:
changed

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+
+const pythonSSLServerFormat = `import http.server, ssl
+
+class RequestHandler(http.server.BaseHTTPRequestHandler):
```
Why not keep it as a separate file?

guyarb replied:
since it is not too big, and it can fit in the file, I left it here for simple modification and changes

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+`
+
+func TestOpenSSLVersions_python_matrix(t *testing.T) {
+	if !httpSupported(t) {
```
Don't we want to make sure it's not enabled there?

guyarb replied:
I'm enabling the feature via configuration later in the function
but i have to check if the kernel is supported

DanielLavie replied:
I mean, do we have a test where the feature is enabled but the kernel is not supported and what is the behaviour?

guyarb replied:
those tests are being run in the kitchen tests, some of the machines there are with older kernels than 4.14
and for those use-cases we just skip the test

guyarb committed:
usm: ssl: openssl: Increased timeouts

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+	go func() {
+		err := cmd.Start()
+		if err != nil {
+			fmt.Println(err)
```
What about the test in this case?

guyarb replied:
https://github.com/DataDog/datadog-agent/pull/13656#discussion_r985251795

DanielLavie replied:
What should happen in the test if this happens? it's an asynchronous operation

guyarb replied:
that's why I'm waiting on the port to be ready

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+	}()
+	defer func() {
+		if cmd.Process != nil {
+			cmd.Process.Kill()
```
What about the test in this case?

guyarb replied:
what?

DanielLavie replied:
What should happen in the test if this happens? it's an asynchronous operation

guyarb replied:
that's a defer

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+	cfg.BPFDebug = true
+	tr, err := NewTracer(cfg)
+	require.NoError(t, err)
+	err = tr.RegisterClient("1")
```
What is "1"?

guyarb replied:
client id, copied from other tests

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+	for {
+		select {
+		case <-ctx.Done():
+			t.Fatalf("failed connecting to port %s:%s", host, port)
```
There's also host

guyarb replied:
changed

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+		default:
+			conn, err := net.DialTimeout("tcp", net.JoinHostPort(host, port), time.Second)
+			if err != nil {
+				continue
```
Why continue?

guyarb replied:
being done in a loop
it is reasonable that it will take us time to open a port, so checking it in intervals
the inbetween errors are legit

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+	statusCodes = []int{nethttp.StatusOK, nethttp.StatusMultipleChoices, nethttp.StatusBadRequest, nethttp.StatusInternalServerError}
+)
+
+func simpleGetRequestsGenerator(t *testing.T, targetAddr string) func() *nethttp.Request {
```
Please be consistent with the location of the helper functions in the file. Also remember the "read a book" rule of thumb

guyarb replied:
moved

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+httpd.serve_forever()
+`
+
+func TestOpenSSLVersions_python_matrix(t *testing.T) {
```
Please add a description for what this test asserts and what are the steps of this test so it will be easier to understand and verify this test

guyarb replied:
added

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+
+	requestFn := simpleGetRequestsGenerator(t, "127.0.0.1:8001")
+	var requests []*nethttp.Request
+	for i := 0; i < 100; i++ {
```
100 -> Magic number

guyarb replied:
changed

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+func assertAllRequestsExists(t *testing.T, tracer *Tracer, requests []*nethttp.Request) {
+	requestsExist := make([]bool, len(requests))
+	for i := 0; i < 10; i++ {
+		time.Sleep(10 * time.Millisecond)
```
Why the code sleeps here?

guyarb replied:
to give time for the kernel to flush the buffers to the usermode

DanielLavie replied:
Please add a comment about it for future us/others

guyarb replied:
added

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+	requestsExist := make([]bool, len(requests))
+	for i := 0; i < 10; i++ {
+		time.Sleep(10 * time.Millisecond)
+		conns, err := tracer.GetActiveConnections("1")
```
"1"?

guyarb replied:
client id, copied from other tests

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,169 @@ func skipIfWindows(t *testing.T) {
+	}
+}
+
+func assertAllRequestsExists(t *testing.T, tracer *Tracer, requests []*nethttp.Request) {
```
Do we also assert al requests exist exactly the number of times they were sent?

guyarb replied:
each is being sent only once

guyarb committed:
usm: ssl: openssl: Fixed CR notes

guyarb committed:
usm: ssl: openssl: Fixed a bug

guyarb committed:
usm: ssl: openssl: Fixed CR notes

guyarb committed:
Added missing check

guyarb committed:
Fixed hanging requests

nplanel commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1883,186 @@ func skipIfWindows(t *testing.T) {
+	numberOfIterations    = 100
+	pythonSSLServerFormat = `import http.server, ssl
+
+class RequestHandler(http.server.BaseHTTPRequestHandler):
```
please move the python server logic in pkg/network/http/testutil/testutil.go with the same API (this would be useful for other test and in http/monitor_test.go)
HTTPPythonServer(t *testing.T, addr string, options Options) func()

guyarb replied:
done

guyarb committed:
Changed configuration initialization to populate runtimeCompiler flag

guyarb committed:
Fixed tests on arm

guyarb committed:
Fixed CR comment

guyarb committed:
Merge branch 'main' into guy.arbitman/USMO-3-python3.10-openssl-support

guyarb committed:
Fixed CI

guyarb committed:
Fixed ci

guyarb committed:
Fixed CI

guyarb committed:
Fixed lint

guyarb committed:
Fixed CI

guyarb committed:
Fixed CI

nplanel commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1881,119 @@ func skipIfWindows(t *testing.T) {
+	}
+
+	client.CloseIdleConnections()
+	assertAllRequestsExists(t, tr, requests)
```
please reuse the same mechanism that we do for https, it's better for visibility when rereading/debugging the test
`require.Eventuallyf() 
``tr.GetActiveConnection() ` it's the same path that we use in sysprobe on the server side (`cmd/system-probe/modules/network_tracer.go`)
https://github.com/DataDog/datadog-agent/blob/main/pkg/network/tracer/tracer_test.go#L1809-L1813

guyarb replied:
done

nplanel commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1881,3 +1881,119 @@ func skipIfWindows(t *testing.T) {
+
+		resp, err := client.Do(req)
+		require.NoError(t, err)
+		require.Equal(t, status, resp.StatusCode)
```
the client will use another tcp connection if you doesn't read/flush the fd, that may why closing took more time as the go http client stack would not close the connections and probably keep alive. (even if you Body.Close())
The normal/expected behaviour from the client would be `io.ReadAll(resp.Body); resp.Body.Close()`

guyarb committed:
Checking improvement to the tests

guyarb committed:
Merge branch 'main' into guy.arbitman/USMO-3-python3.10-openssl-support

guyarb committed:
restored code

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1922,104 +1896,55 @@ func TestOpenSSLVersions(t *testing.T) {
+	cfg := testConfig()
 	cfg.EnableHTTPSMonitoring = true
 	cfg.EnableHTTPMonitoring = true
+	cfg.BPFDebug = true
```
Why?

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1922,104 +1896,55 @@ func TestOpenSSLVersions(t *testing.T) {
-	portCtx, cancelPortCtx := context.WithDeadline(context.Background(), time.Now().Add(time.Second*5))
-	rawConnect(portCtx, t, "127.0.0.1", "8001")
-	cancelPortCtx()
+	addr := "127.0.0.1:8001"
```
address of what?

DanielLavie commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -2055,7 +1983,7 @@ func assertAllRequestsExists(t *testing.T, tracer *Tracer, requests []*nethttp.R
 	// If we reach here, it means that at least one req is missing, so we want to tell who is missing.
 	for reqIndex, exists := range requestsExist {
-		require.Truef(t, exists, "request %d was not found (req %v)", reqIndex, requests[reqIndex])
+		require.Truef(t, exists, "request %d was not found (req %v)", reqIndex+1, requests[reqIndex])
```
Why reqIndex+1? please add a comment

guyarb replied:
added a comment

guyarb committed:
Fixed cr comments

guyarb committed:
Fixed cr comments

guyarb committed:
Fixed cr comments
